core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java

```diff
@@ -837,6 +841,44 @@ private static RexNode makeConstructorCall(
     return rexBuilder.makeNewInvocation(type, defaultCasts);
   }
 
+  private RexNode convertItem(
+      @UnknownInitialization StandardConvertletTable this,
+      SqlRexContext cx,
+      SqlCall call) {
+    assert call.operandCount() == 2;
```

Comment:
What ensures this assert never fails?
What happens in case user types `x[2, 3, 4]` in SQL? Should the error be user-friendly?

Reply:
It's the parser which prevents multiple operands for ITEM, it expects a single expression/operand between the brackets ([Parser.jj#L3576](https://github.com/apache/calcite/blob/0ab1fe3118cc5adb8750a5c785e24fa4381afda1/core/src/main/codegen/templates/Parser.jj#L3576
))

The preceding phases, type checking ([SqlItemOperator.java#L89](https://github.com/apache/calcite/blob/0ab1fe3118cc5adb8750a5c785e24fa4381afda1/core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java#L89)) for operands and return type inference ([SqlItemOperator.java#L144](https://github.com/apache/calcite/blob/0ab1fe3118cc5adb8750a5c785e24fa4381afda1/core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java#L144)) for ITEM, already rely on having a single operand (in addition to the implicit one).

Reply:
Should the assert be removed then?

Reply:
I think you are right and the assertion is redundant, the other preceding steps discussed in a previous comment would already fail with different number of operands.

I added it in the first place mostly for consistency sake with the rest of the methods of the same file, so I am fine with dropping it (and very much true that, as-is, it would not make the life easier for investigating failures).